walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,102 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+
+    // first iterate through and combine events that overlap
+    // subtract combined events calendar from whole day
+    // remove time ranges that are too short 
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+        List<TimeRange> avaliableRanges = new ArrayList<TimeRange>();
+        avaliableRanges.add(TimeRange.WHOLE_DAY);
+        Collection<TimeRange> bookedRanges = combinedMandatoryRanges(events, request);
+
+        for (TimeRange booked : bookedRanges) {
+            TimeRange lastInAvaliable = avaliableRanges.get(avaliableRanges.size()-1);
+            if (lastInAvaliable.overlaps(booked)){
+                List<TimeRange> newLastAvaliable = removeOverlap(lastInAvaliable, booked);
+                avaliableRanges.remove(lastInAvaliable);
+                avaliableRanges.addAll(newLastAvaliable);
+                Collections.sort(avaliableRanges, TimeRange.ORDER_BY_START);
+            }
+        }
+
+        // remove TimeRanges too small
+        for (TimeRange avaliable : avaliableRanges) {
+            if (avaliable.duration() < request.getDuration()) {
+                avaliableRanges.remove(avaliable);
+            }
+        }
+
+        return avaliableRanges;
+  }
+
+// returns set of TimeRanges formed by all events that have mandatory attendees from the request with no overlap
+  public Collection<TimeRange> combinedMandatoryRanges(Collection<Event> events, MeetingRequest request) {
+        List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+        for (Event event : events) {
+            if (!Collections.disjoint(event.getAttendees(), request.getAttendees()) ) { //events with mandatory attendees only
+                busyTimes.add(event.getWhen());
+                System.out.println(event.getWhen());
```

Comment:
Avoid leaving debugging statements in code. In production systems, you'd use a logging library, like java.util.logging.Logger or something that uses that, where you can specify different logging levels (debug, info, warning, severe), and then you can configure what level of logs you want to see. In this case, you'd do something like logger.debug(event.getWhen()), so if someone's debugging this class, they can turn on printing out debug logs and see this line, and not see this information otherwise

Reply:
Thanks